core/src/main/scala/kafka/log/TimeIndex.scala

```diff
@@ -240,24 +241,34 @@ object TimeIndex extends Logging {
   */
 class LazyTimeIndex(@volatile private var _file: File, baseOffset: Long, maxIndexSize: Int = -1, writable: Boolean = true) {
   @volatile private var timeIndex: Option[TimeIndex] = None
+  private val lock = new ReentrantLock()
 
   def file: File = {
-    if (timeIndex.isDefined)
-      timeIndex.get.file
-    else
-      _file
+    inLock(lock) {
```

Comment:
i think this lock is useless because this is a get file method.if modify the file, also can get correct file at next time.

Reply:
I believe getter is also susceptible for race. Consider the following situation:
 - Thread 1 enters "else" codepath of the `file` getter and is ready to return the reference of `_file`
 - At this very moment Thread 2 performs two operations consecutively: calls `get` and immediately `file_=` setter replacing the current `file` reference within `TimeIndex` (but leaving `LazyTimeIndex`'s `_file` outdated
 - Thread 1 continues its operation and returns a stale reference to the File object which is no longer correct

Reply:
I think a better way to handle this is to replace `Option[TimeIndex]` with two case classes that inherit from a common sealed trait. Then we don't have this issue where we have two keep 2 separate fields in sync.

Reply:
@ijuma agreed, please check out the latest commit